src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp

```diff
@@ -510,6 +521,23 @@ void ShenandoahConcurrentGC::entry_evacuate() {
   op_evacuate();
 }
 
+void ShenandoahConcurrentGC::entry_promote_in_place() {
+  shenandoah_assert_generational();
+
```

Comment:
Do you also want to assert that collection set is empty?

Reply:
The other approach might be to _always_ have a separate pre-evac phase that promotes in place before actual evac to follow (even when collection set is non-empty).

Reply:
Yes, added this assertion in f409f645e1.